include/kernels/gauge_force.cuh

```diff
@@ -14,47 +14,57 @@ namespace quda {
     int *input_path[4];
     const int *length;
     const double *path_coeff;
+    int *buffer;
     int count;
 
-    paths(void *buffer, size_t bytes, int ***input_path, int *length_h, double *path_coeff_h, int num_paths, int max_length) :
+    paths(int ***input_path, int *length_h, double *path_coeff_h, int num_paths, int max_length) :
       num_paths(num_paths),
       max_length(max_length),
       count(0)
     {
-      void *path_h = safe_malloc(bytes);
+      // create path struct in a single allocation
+      size_t bytes = 4 * num_paths * max_length * sizeof(int) + num_paths * sizeof(int);
+      int pad = ((sizeof(double) - bytes % sizeof(double)) % sizeof(double))/sizeof(int);
+      bytes += pad*sizeof(int) + num_paths*sizeof(double);
+
+      buffer = static_cast<int*>(pool_device_malloc(bytes));
+      int *path_h = static_cast<int*>(safe_malloc(bytes));
       memset(path_h, 0, bytes);
 
-      int *input_path_h = (int*)path_h;
       for (int dir=0; dir<4; dir++) {
         // flatten the input_path array for copying to the device
         for (int i=0; i < num_paths; i++) {
           for (int j=0; j < length_h[i]; j++) {
-            input_path_h[dir*num_paths*max_length + i*max_length + j] = input_path[dir][i][j];
+            path_h[dir*num_paths*max_length + i*max_length + j] = input_path[dir][i][j];
             if (dir==0) count++;
           }
         }
       }
 
       // length array
-      memcpy((char*)path_h + 4 * num_paths * max_length * sizeof(int), length_h, num_paths*sizeof(int));
+      memcpy(path_h + 4 * num_paths * max_length, length_h, num_paths*sizeof(int));
 
       // path_coeff array
-      memcpy((char*)path_h + 4 * num_paths * max_length * sizeof(int) + num_paths*sizeof(int), path_coeff_h, num_paths*sizeof(double));
+      memcpy(path_h + 4 * num_paths * max_length + num_paths + pad, path_coeff_h, num_paths*sizeof(double));
 
       qudaMemcpy(buffer, path_h, bytes, qudaMemcpyHostToDevice);
       host_free(path_h);
 
       // finally set the pointers to the correct offsets in the buffer
-      for (int d=0; d < 4; d++) this->input_path[d] = (int*)((char*)buffer + d*num_paths*max_length*sizeof(int));
-      length = (int*)((char*)buffer + 4*num_paths*max_length*sizeof(int));
-      path_coeff = (double*)((char*)buffer + 4 * num_paths * max_length * sizeof(int) + num_paths*sizeof(int));
+      for (int d=0; d < 4; d++) this->input_path[d] = buffer + d*num_paths*max_length;
+      length = buffer + 4*num_paths*max_length;
+      path_coeff = reinterpret_cast<double*>(buffer + 4 * num_paths * max_length + num_paths + pad);
+    }
+    void free() {
```

Comment:
I know I could use a destructor, but the class is copied at some point. I also tried using `shared_ptr` but without success...